readme.md

```diff
@@ -47,126 +62,497 @@ This package exports [ESM](https://gist.github.com/sindresorhus/a39789f98801d908
+
+   ```typescript
+   import { ChatModel } from '@dexaai/dexter';
+
+   const customCache = new Map();
+   const chatModel = new ChatModel({ cache: customCache });
+   ```
+
+4. **Cache Usage**: Caching is automatically applied in the `AbstractModel` class. Before making an API call, it checks the cache for a stored response. After receiving a response, it stores it in the cache for future use.
+
+5. **Cache Invalidation**: Cache invalidation is left to the user. You can implement your own cache invalidation strategy based on your specific use case.
+
+### Tokenization
+
+Dexter includes a tokenization system based on the `tiktoken` library, which is used by OpenAI for their models. This system is crucial for accurately counting tokens and managing model inputs and outputs.
+
+1. **Tokenizer Creation**: The `createTokenizer` function creates a `Tokenizer` instance for a specific model:
+
+   ```typescript
+   const tokenizer = createTokenizer('gpt-3.5-turbo');
+   ```
+
+2. **Tokenizer Methods**:
+   - `encode(text: string): Uint32Array`: Encodes text to tokens
+   - `decode(tokens: number[] | Uint32Array): string`: Decodes tokens to text
+   - `countTokens(input?: string | ChatMessage | ChatMessage[]): number`: Counts tokens in various input formats
+   - `truncate({ text: string, max: number, from?: 'start' | 'end' }): string`: Truncates text to a maximum number of tokens
+
+3. **Model Integration**: Each model instance has its own `Tokenizer`, which is used internally for token counting and management.
+
+### Model Event Hooks
+
+Dexter provides a system of event hooks that allow you to add custom logic at various points in the model execution process. These hooks are defined in the `Model.Events` interface:
+
+1. **Available Hooks**:
+   - `onStart`: Called before the model execution starts
+   - `onApiResponse`: Called after receiving a response from the API
+   - `onComplete`: Called after the model execution is complete
+   - `onError`: Called if an error occurs during model execution
+
+2. **Hook Parameters**: Each hook receives an event object with relevant information, such as timestamps, model parameters, responses, and context.
+
+3. **Usage**: Event hooks can be defined when creating a model instance:
+
+   ```typescript
+   const chatModel = new ChatModel({
+     events: {
+       onStart: [(event) => console.log('Starting model execution', event)],
+       onComplete: [(event) => console.log('Model execution complete', event)],
+     },
+   });
+   ```
+
+4. **Multiple Handlers**: Each event can have multiple handlers, which are executed in the order they are defined.
+
+5. **Async Handlers**: Event handlers can be asynchronous functions. Dexter uses `Promise.allSettled` to handle multiple async handlers.
+
+6. **Extending Models**: When using the `extend` method to create a new model instance, event handlers are merged, allowing you to add new handlers without removing existing ones.
```

Comment:
`.extend` is sexy; we should add an example of this in the future.

Reply:
In that example, show something useful it can be used for.